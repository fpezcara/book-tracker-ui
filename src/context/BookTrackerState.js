import React, { useReducer, useEffect, useMemo, useState } from "react";
import BookTrackerContext from "./book-tracker-context";
import BookTrackerReducer from "./book-tracker-reducer";

import { API_URL } from "../utils/constants";
import Cookies from "js-cookie";

import {
  ADD_BOOK,
  DELETE_BOOK,
  CURRENT_BOOKLIST,
} from "./book-tracker-actions";

const BookTrackerState = ({ children }) => {
  const { Provider } = BookTrackerContext;
  const userId = Cookies.get("userId");
  // todo: rename to bookLists
  const [lists, setLists] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState({});

  const updateCurrentBookList = (selectedBookList) => {
    dispatch({
      type: CURRENT_BOOKLIST,
      payload: selectedBookList,
    });
  };

  const addBook = (book) => {
    dispatch({
      type: ADD_BOOK,
      payload: book,
    });
  };
  const deleteBook = (book) => {
    dispatch({
      type: DELETE_BOOK,
      payload: book,
    });
  };
  const initialState = useMemo(() => {
    return {
      bookLists: [
        { id: 1, listUrl: "reading", listName: "Reading", books: [] },
        { id: 2, listUrl: "finished", listName: "Finished", books: [] },
        { id: 3, listUrl: "wishlist", listName: "Wishlist", books: [] },
      ],
      currentBookList: "reading",
    };
  }, []);

  const [state, dispatch] = useReducer(BookTrackerReducer, initialState);

  useEffect(() => {
    if (JSON.parse(localStorage.getItem("state"))) {
      dispatch({
        type: "init_stored",
        payload: JSON.parse(localStorage.getItem("state")),
      });
    }
  }, []);

  useEffect(() => {
    try {
      userId &&
        fetch(`${API_URL}/users/${userId}/lists`, {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
          },
          credentials: "include",
        })
          .then((response) => response.json())
          .then((data) => {
            console.log("Fetched data:", data);
            setLists(data);
            setLoading(false);
          });
    } catch (error) {
      console.error("Error fetching book lists:", error);
      setError(error);
      setLoading(false);
    }

    if (state !== initialState) {
      localStorage.setItem("state", JSON.stringify(state));
    }
  }, [state, initialState, userId]);

  const values = {
    state,
    addBook,
    deleteBook,
    updateCurrentBookList,
    lists,
    loading,
    error,
  };

  return <Provider value={values}>{children}</Provider>;
};

export default BookTrackerState;
