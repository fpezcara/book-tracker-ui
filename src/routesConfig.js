import Home from "./components/Home";
import Header from "./components/Header";
import AddBook from "./components/AddBook";
import NotFound from "./components/NotFound";
import Registration from "./components/Registration";
import Login from "./components/Login";

import { Outlet, Navigate } from "react-router-dom";

const routesConfig = [
  {
    path: "/",
    element: (
      <>
        <Header />
        <Outlet />
      </>
    ),
    errorElement: <NotFound />,
    children: [
      { path: "/", element: <Navigate to="/reading" /> },
      { path: "/404", element: <NotFound /> },
      { path: "/:name", element: <Home /> },
      { path: "/:name/add-book", element: <AddBook /> },
      { path: "/register", element: <Registration /> },
      { path: "/login", element: <Login /> },
    ],
  },
];

export default routesConfig;
