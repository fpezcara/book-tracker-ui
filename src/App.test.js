import { render, screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { RouterProvider, createMemoryRouter } from "react-router";
import routesConfig from "./routesConfig";
import BookTrackerState from "./context/BookTrackerState";
import Cookies from "js-cookie";
import axios from "axios";

const mockBookLists = [
  {
    id: "dd578ff0-bfdb-4f16-83ae-3d9f9aa661c4",
    name: "Reading",
    updated_at: "2025-03-23T15:26:58.317Z",
    user_id: 9,
    created_at: "2025-03-23T15:26:58.317Z",
    books: [],
  },
  {
    id: "dd578ff0-b0eb-4f16-83ae-3d9f9aa661c4",
    name: "Wishlist",
    updated_at: "2025-03-23T15:26:58.317Z",
    user_id: 9,
    created_at: "2025-03-23T15:26:58.317Z",
    books: [],
  },
  {
    id: "dd578ff0-bfdb-4g16-83ae-3d9f9aa761c4",
    name: "Finished",
    updated_at: "2025-03-23T15:26:58.317Z",
    user_id: 9,
    created_at: "2025-03-23T15:26:58.317Z",
    books: [],
  },
];

jest.mock("axios");

describe("App", () => {
  describe("When user is logged out", () => {
    test("redirects to /login", () => {
      const router = createMemoryRouter(routesConfig, {
        initialEntries: ["/reading"],
      });

      render(
        <BookTrackerState>
          <RouterProvider router={router} />
        </BookTrackerState>,
      );

      expect(router.state.location.pathname).toBe("/login");
    });
  });

  describe("When user is logged in", () => {
    beforeEach(() => {
      Cookies.set("userId", "9");

      axios.get.mockResolvedValue({ data: mockBookLists });
    });

    afterEach(() => {
      Cookies.remove("userId");
      Cookies.remove("currentBookList");

      jest.clearAllMocks();
    });

    test("redirects to /reading on first render", async () => {
      const router = createMemoryRouter(routesConfig, {
        initialEntries: ["/"],
      });
      Cookies.set("currentBookList", "reading");

      const { container } = render(
        <BookTrackerState>
          <RouterProvider router={router} />
        </BookTrackerState>,
      );

      await waitFor(() => {
        screen.queryByTestId("booklist-select");
      });

      const bookTrackerTitle = await screen.findByRole("heading", { level: 1 });
      const bookList = await screen.findByRole("heading", { level: 3 });

      expect(bookTrackerTitle).toHaveTextContent(/book tracker/i);
      expect(bookList).toHaveTextContent(/reading/i);
      expect(screen.getByText(/no books have been added/i)).toBeInTheDocument();
      expect(router.state.location.pathname).toBe("/reading");
      expect(container).toMatchSnapshot();
    });

    test("selecting a booklist from the select input navigates to the right booklist name", async () => {
      const router = createMemoryRouter(routesConfig, {
        initialEntries: ["/reading"],
      });
      Cookies.set("currentBookList", "reading");

      render(
        <BookTrackerState>
          <RouterProvider router={router} />
        </BookTrackerState>,
      );

      await waitFor(() => {
        expect(screen.getByTestId("booklist-select")).toBeInTheDocument();
      });

      await screen.findByText("Reading");

      const booklistSelect = screen.queryByTestId("booklist-select");
      const bookListHeading = screen.queryByRole("heading", { level: 3 });

      await userEvent.selectOptions(booklistSelect, "finished");

      expect(Cookies.get("currentBookList")).toBe("finished");

      expect(bookListHeading).toHaveTextContent(/finished/i);

      expect(booklistSelect.value).toBe("finished");
    });

    test("shows page not found when page does not exist", async () => {
      const router = createMemoryRouter(routesConfig, {
        initialEntries: ["/random"],
      });

      render(
        <BookTrackerState>
          <RouterProvider router={router} />
        </BookTrackerState>,
      );

      const goHomeText = await screen.findByText(/go home/i);

      expect(goHomeText).toBeInTheDocument();
      expect(screen.getByAltText("page not found")).toBeVisible();
    });

    test("clicking on Add Book button navigates to /add-book", async () => {
      const router = createMemoryRouter(routesConfig, {
        initialEntries: ["/finished"],
      });
      Cookies.set("currentBookList", "finished");

      render(
        <BookTrackerState>
          <RouterProvider router={router} />
        </BookTrackerState>,
      );

      const addBookButton = await screen.findByTestId("add-book-button");
      userEvent.click(addBookButton);

      expect(await screen.findByTestId("search-by-input")).toBeInTheDocument();
      expect(
        await screen.findByTestId("search-by-options-select"),
      ).toBeInTheDocument();
    });

    test("clicking on Home button when in /add-book redirects to main page", async () => {
      const router = createMemoryRouter(routesConfig, {
        initialEntries: ["/wishlist", "/wishlist/add-book"],
      });
      Cookies.set("currentBookList", "wishlist");

      render(
        <BookTrackerState>
          <RouterProvider router={router} />
        </BookTrackerState>,
      );
      Cookies.set("currentBookList", "wishlist");

      await waitFor(() => {
        expect(
          screen.getByTestId("search-by-options-select"),
        ).toBeInTheDocument();
      });

      const homeButton = await screen.findByText("Home");

      userEvent.click(homeButton);

      expect(await screen.findByTestId("home-container")).toBeInTheDocument();
      await waitFor(() => {
        expect(
          screen.getByText(/no books have been added/i),
        ).toBeInTheDocument();
      });
      expect(router.state.location.pathname).toBe("/wishlist");
    });
  });
});
