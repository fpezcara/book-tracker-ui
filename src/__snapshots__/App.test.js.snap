// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App When user is logged in redirects to /reading on first render 1`] = `
<div>
  <header
    class="sc-jOTjWM hMgtXy"
  >
    <div
      class="authentication"
    >
      <a
        class="link home"
        data-discover="true"
        href="/"
      >
        Home
      </a>
      <div
        class="authLinks"
      >
        <a
          class="link"
          data-discover="true"
          href="/logout"
        >
          Logout
        </a>
      </div>
    </div>
    <div>
      <h1>
        Book Tracker
      </h1>
    </div>
  </header>
  <main
    class="sc-gnOvAp iUFPTw"
    data-testid="home-container"
  >
    <div
      class="sc-dQkurY bgnaKH"
    >
      <div
        class="sc-glJfXD jVIZTa"
      >
        <h3>
          reading
        </h3>
        <select
          data-testid="booklist-select"
        >
          <option
            value="reading"
          >
            Reading
          </option>
          <option
            value="wishlist"
          >
            Wishlist
          </option>
          <option
            value="finished"
          >
            Finished
          </option>
        </select>
      </div>
      <div
        class="sc-dOfePm pRbiN"
      >
        <button
          class="sc-VHjGu fcYvPT add-book"
          data-testid="add-book-button"
        >
          Add book
        </button>
      </div>
    </div>
    <table
      class="sc-jgmqJh gcmQwT"
    >
      <thead>
        <tr>
          <th>
            Cover
          </th>
          <th>
            Title
          </th>
          <th>
            Author
          </th>
          <th>
            Pages
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            colspan="4"
          >
            No books have been added.
          </td>
        </tr>
      </tbody>
    </table>
  </main>
</div>
`;
